src/types.ts

```diff
@@ -36,3 +36,9 @@ type ListenerObject<Ev extends Event> = {
   // For listener objects, the handleEvent function has the object as the `this` binding
   handleEvent(this: ListenerObject<Ev>, Ee: Ev): void;
 };
+
+export type PossibleEventType<K> = K extends any
+  ? K extends `on${infer Type}`
+    ? Type
+    : never
```

Comment:
what would happen for an event that did not have a on* property (eg `DOMContentLoaded`). Would that be allowed, or would it be an error now?

Reply:
Should it be `string` rather than `never`?

Reply:
Ideally we get auto complete + inference goodness. But when the event has no `on*` it just gets an `Event`

Reply:
> Should it be string rather than never?

This might turn off the autocomplete etc as we basically distribute a union of **literal** strings here and narrow it down to a union of **literal** strings (the operation here acts as a `.filter()`). We can't remap this union to smth like `'foo' | string` because that's basically `string` (types reduce themselves to a common denominator).

> But when the event has no on* it just gets an Event

Could you give me a practical example of when this would be useful? 

> what would happen for an event that did not have a on* property (eg DOMContentLoaded). Would that be allowed, or would it be an error now?

Most likely an error now - but since this special case is known upfront I could try to introduce support for it.

---

There is also a chance that maybe this could be actually implemented with an additional overload. This is how `window.addEventListener` is implemented:
```ts
    addEventListener<K extends keyof WindowEventMap>(type: K, listener: (this: Window, ev: WindowEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
```
And it comes with autocomplete and allows u to use a random string.

As mentioned elsewhere - I'm not a huge fan of overloads as they don't play that well with some TS features (for instance, when using helpers like `ReturnType` or a conditional type with `infer` then only the very last overload is taken into a consideration and all the previous ones are ignored). This could work OK in this context here though.

Reply:
Fwiw I liked #10 that it handled all event targets.

(I guess eventually I'll probably make an events package in [sthir](https://github.com/devanshj/sthir) that would only have types required to deal with event emitters, it's a common problem eg even RxJS's `fromEvent` could leverage it)